### ISSUE TITLE
Add example of gtk 4 GtkListVIew

### ISSUE THREAD

**Issue by gavr123456789** [StefanSalewski/gintro#126]
![image](https://user-images.githubusercontent.com/30507409/119156626-14e1cf80-ba5d-11eb-8a19-a7119f4ba1ba.png)
https://blog.gtk.org/2020/06/07/scalable-lists-in-gtk-4/  
  
  
This is the most productive of all containers, and judging by the guides, its implementation is also the most difficult.   
https://github.com/ToshioCP/Gtk4-tutorial/blob/main/gfm/sec24.md    
https://github.com/ToshioCP/Gtk4-tutorial/blob/main/gfm/sec25.md

**Comment by StefanSalewski** [comment]
Yes I know the gtk4 tutorial of Mr Toshio, he announced it a few months ago. I read a few of his early explanations, that was not too good, but maybe he has improved that now.

Explaining ListView/Treeview in a manner that people can understand and remember it easily is unfortunately not easy. In the book of A. Krause that took already 30 pages. I may continue with that in winter.

Currently my focus is more the Nim beginner book, I have about 200 pages ready now. That book has at least 16 stars at github and was mentioned recently at reddit with 50 upvotes, so I assume it has some relevance.

Unfortunately GTK4 has not attracted that many new users generally, and my feeling is that the number of gintro users is still close to zero after all these years. And the Nim guys seems to promote figma now. Well that treeform guy puts indeed much effort into figma, we will see if it has some success.

Have you made progress with adwaita and gtksourceview? GTK4 became recently available for gentoo, so I may continue with some work on my gtksourceview editor soon, I already pushed it to github.

**Comment by gavr123456789** [comment]
okay, here's my journey on this 20 days.   
I have a [project](https://github.com/gavr123456789/FileBrowser) on vala, which was not completed due to the fact that I changed jobs. This is a file browser with the idea of management without pcm. I also planned to preview things like images, audio, and video in the form of opening new carousel pages, going inside tree documents(json, xml, sqlite) as if they were still folders and files.  
  
The project was started on gtk 3, and I had to manually do optimizations like [lazy loading](https://github.com/gavr123456789/FileBrowser/blob/master/src/Views/Page.vala#L39) due to the poor performance of gtk 3.   
Now that I have some free time I started the implementation on nim with the console version, but then I realized that it's easier to start with the gui right away. Next, I ran into a problem - how to find out which page of the carousel was clicked inside the activation callback of one of the entries.  
  
In the project on vala, I did this by sending signals through the entire hierarchy, from record to page, from page to main widgets, so I knew which record, from which page was clicked [1](https://github.com/gavr123456789/FileBrowser/blob/master/src/Views/RowWidget.vala#L75) [2](https://github.com/gavr123456789/FileBrowser/blob/master/src/Views/Page.vala#L116).  
  
On nim, I decided to abandon this idea because I want to try to make a project without inheritance, and I also did not find a guide on how to create custom signals, in vala it is simply built into the language [syntax](https://github.com/gavr123456789/FileBrowser/blob/master/src/Views/Page.vala#L22).   
  
I came up with three options, the first is to store the page number data directly in the records via setData, this is a bad option since typing is lost and I'm not sure about memory managent, since setData takes a pointer. The second option is to store the record models as a singleton structure with a hash table ListBox - > Model. I also added an example of this to my [PR](https://github.com/gavr123456789/gintro/blob/add_examples/examples/gtk4/hash_tables.nim).
  
I came to the third one only recently, I don't know GTK 4 very well yet, I'm trying it with Nim for the first time. Actually, this is the obvious option that I should have started with - using ListView with ListModel. So first I need to rewrite the examples from tutorial of Mr. Toshio from C to Nim.  
  
> Have you made progress with adwaita  
  
Yes https://github.com/gavr123456789/gintro/blob/add_examples/examples/gtk4/adwaita/carousel.nim.   
For now I tried adwaita application widnow(the bottom edges of the app are rounded as well as the upper ones), [adw.HeaderBar](https://github.com/gavr123456789/Katana/blob/master/src/view/MainWindow.nim#L20) and Carousel 

> and gtksourceview  
  
No, So far, it seemed to me that it is very complex. I'll need it at a later stage when I'm done with creating a regular file browser, and I'll need the ability to edit when previewing.   
My overly ambitious idea is to create something like an object browser from Smalltalk, but for code in other languages, and also combine this with a file/TreeStructureFiles browser. I found that someone has already done something similar, but on Qt https://tibleiz.net/code-browser/.

**Comment by gavr123456789** [comment]
While I was creating [this](https://github.com/gavr123456789/GTK-info/blob/master/SHOWCASE.MD) list of applications that got into [GNOME Circle](https://circle.gnome.org/), I noticed that there is [one application](https://flathub.org/apps/details/dev.geopjr.Hashbrown) in Crystal (it only has bindings to GTK 3), an even more unpopular language than Nim. I think maybe its a good idea to write a small application and also put it there for populizing Nim + GTK. People just don't know it exists. GTK devs have not heard about the existence of Nim, and if they have, they do not know that it can be used with GTK. At least that's the impression I got.

**Comment by StefanSalewski** [comment]
>GTK devs have not heard about the existence of Nim

Meanwhile many people have heard of Nim. The situation was a bit different in 2014 when I discovered Nim, that time when I talk with people they had no idea what that Nimrod is. Today a lot of people heard about it already, some even tried. But many have reasons not continuing using it. Unfortunately a lot of bright people left Nim community in the last 5 years. Gtk people know Nim, I mentioned it a few times on the GTK forum, and the GTK4 book was on reddit recently. But who uses GTK at all currently? A few Rust people. Have you seen new GTK projects in the last 10 years? Well there are a few done with Rust. In the past people complained that GTK has not enough tutorials, but my books or the D tutorial of Don have not really improved the situation. For Crystal, long time it was more popular than Nim as many Ruby users went from Ruby to Crystal. For your Vala stuff -- I had the same problem with Ruby: Did a schematic editor and a PCB router in Ruby, see  http://ssalewski.de/PetEd.html.en and http://ssalewski.de/Router.html.en. Unfortunately it took me about 1600 hours to get the gintro bindings to a useful state, so I had not much time to work on the Ruby projects in the last seven years. But recently I ported the router, see https://github.com/StefanSalewski/RBR. That was easy, took me only 100 hours, and I like the result. For porting the schematic editor from Ruby to Nim I got it compiling after maybe 200 hours, but as the Ruby version had classes, and now I intend to use Gtk4 I finally wrote all from scratch. And I intend to do a more general CAD tool now. Initially it was all closely related to the gEDA project, but as that is nearly death now it makes no sense to be compatible in any way. The new Nim CAD tool is also already at github, but in an very early state.
